src/programs/md5/digest.1
.\" $NetBSD: digest.1,v 1.1 2002/12/21 03:41:00 schmonz Exp $
.\"
.\"
.\" Copyright (c) 2001 Alistair G. Crooks.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by Alistair G. Crooks.
.\" 4. The name of the author may not be used to endorse or promote
.\"    products derived from this software without specific prior written
.\"    permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS
.\" OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
.\" DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
.\" GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\" WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
.\" NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
.\" SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\"
.Dd February 28, 2001
.Dt DIGEST 1
.Os
.Sh NAME
.Nm digest
.Nd calculate message digests
.Sh SYNOPSIS
.Nm
.Ar algorithm
.Op file ...
.Sh DESCRIPTION
The
.Nm
utility calculates message digests of files or,
if no file is specified, standard input.
The list of possible algorithms is:
.Bl -tag -width Ds
.It md5
the
.Xr md5 3
algorithm will be used.
.It rmd160
the
.Xr rmd160 3
algorithm will be used.
.It sha1
the
.Xr sha1 3
algorithm will be used.
.It sha256
the
sha256
algorithm will be used.
.It sha384
the
sha384
algorithm will be used.
.It sha512
the
sha512
algorithm will be used.
.El
.Pp
The
.Nm
utility is a simple wrapper for the various different
algorithm implementations, which are located in the standard
C library, and was designed to be scalable as new message digest
algorithms are developed.
.Pp
The
.Nm
utility exits 0 on success, and >0 if an error occurs.
.Sh SEE ALSO
.Xr cksum 1 ,
.Xr md5 3 ,
.Xr rmd160 3 ,
.Xr sha1 3
.Sh HISTORY
The
.Nm
utility first appeared in
.Nx 1.6 .
.Sh AUTHORS
The
.Nm
utility was written by
.An Alistair G. Crooks Aq agc@netbsd.org .


NEWS
== Overview of changes from MacPorts 1.8.2 to 1.9.0 ==

# For regular users:

*) All output is now logged to a file, whose location will be shown if a port
   target fails. This should be attached to bug reports. The new 'log' action
   can also be used to view the log for a port.
*) Command line options can be abbreviated if unambigious.
*) New actions: rdeps, rdependents, setrequested, unsetrequested
*) --follow-dependencies option for uninstall (ports with the requested flag
   set or with other dependents are skipped).
*) New pseudo-ports and selectors: requested, unrequested, leaves, depof:,
   rdepof:, dependentof:, rdependentof:.
*) Architectures of dependencies are checked before install, and in many cases
   will be automatically fixed by upgrading with +universal.
*) New sqlite registry format which is faster and lacks some of the old
   format's inconvenient limitations.
*) Explicitly deselected variants (e.g. -foo) are now recorded, so they will
   not be set to + when upgrading if +foo appears in variants.conf or
   the port's default_variants.

# For Portfile writers:

*) New os.subplatform variable.
*) New configure.ld_archflags option to match configure.cc_archflags et al.
*) 'platform' acts like an 'if' statement rather than a variant.
*) New supported_archs option for indicating a port only supports certain
   archs or is noarch.
*) pre-/post-deactivate and pre-/post-uninstall procedures can now be used,
   and pre-/post-activate procedures are run every time the port is activated,
   not just the first time when it is installed.

== Overview of changes from MacPorts 1.7.1 to 1.8.0 ==

# For regular users:

*) Compatible with Mac OS X 10.6 Snow Leopard, requires at least 10.4 Tiger
*) CPU architecture to build for is now configurable
*) Performance and memory usage improvements
*) New -y option for dry-run
*) 'port search' has many new options that make it more flexible
*) 'port upgrade' now follows the right dependencies when variants change them
*) 'port info' and 'port variants' indicate selected and default variants;
   'port variants' also shows conflicts and dependencies between variants
*) A port can be marked as being replaced by another port; upgrading will
   perform the replacement
*) upgrade accepts new options: --force, --enforce-variants, --no-replace
*) Ports can be marked as conflicting, and conflicts will be detected before
   the install process starts
*) Proxy information can now be obtained from System Preferences and from
   macports.conf
*) Activating a port will now first deactivate any other version of the same
   port that is already active
*) MacPorts can be configured to drop root privileges when not needed
*) port(1) will exit with an appropriate status code when an error is
   encountered (use -p to continue despite errors)

# For Portfile writers:

*) New options: replaced_by, conflicts, depends_fetch, depends_extract, license
*) build_arch variable indicates which architecture should be targeted. Flags
   used to implement this are configured per-compiler, for example,
   configure.cc_archflags will be passed to configure.cc in CFLAGS
*) Dependencies can be satisfied by ports that are installed but not
   available in the ports tree
*) Deleting all the contents of an option no longer unsets the option entirely
*) Using an alternative fetch or extract type now adds appropriate
   dependencies automatically, e.g. 'use_zip yes' adds a dep on unzip, and
   'fetch.type svn' adds a dep on subversion
*) Running phases prior to install should now work reliably when the port is
   already installed

== Overview of changes from MacPorts 1.6.0 to 1.7.0 ==

# For regular users:

*) The resources directory (containing port group information and the mirror sites) has been moved into the ports tree, so a new MacPorts release is no longer needed to update these items
*) The error message from 'port upgrade outdated' when no ports are outdated is much improved
*) Improvement on handling of MacPorts configuration files (global and user-defined options)
*) Case-insensitive issues have been fixed
*) 'port help' is improved
*) The MacPorts mirror, distfiles.macports.org, is automatically checked when the normal sites fail to respond
*) The search command of port now searches the name, description, and long_description fields
*) Numerous improvements to dependency handling
*) Work around a bug in 10.5's Tcl environment handling
*) Fetching distfiles now tries to ping all mirror sites first in an attempt to pick the best source
*) A new option was added to uninstall: --recursive
*) The output for the port commands info, deps, dependents, and search has been improved
*) Two new options to selfupdate: --pretend and --nosync
*) A platform command has been added to port to display your platform information
*) A suitable message is output when the requested port is not installed when polling its dependents;
*) Fixed the "Error: Unable to open port: invalid command name "ui_prefix"" runtime error on Panther;
*) Fixed the "Error: Unable to open port: invalid command name "ui_channels"" runtime error on Panther;
*) Fixed a MacPorts, Panther specific build failure due to missing sqlite headers on that platform;
*) Fixed a MacPorts build failure when building itself in universal fashion;


# For Portfile writers:

*) ruby port group adds a new parameter to ruby.setup for multiple ruby version support
*) New use_dmg option allows DMG-format distfiles
*) ruby.setup (part of the ruby port group) now takes a fetch type
*) git has been added as a valid fetch.type
*) On Mac OS X, MACOSX_DEPLOYMENT_TARGET is now always set to workaround various issues
*) Several universal-specific variables added to the configure.* group
*) A merge function is now available for easier merging of multiple architectures for universal support
*) Pre-fabricated statements with correct checksums are output when in verbose mode if the existing ones in the Portfile are incorrect;

src/programs/md5/digest.c
/*	$NetBSD: digest.c,v 1.1.1.1 2002/09/19 10:44:28 agc Exp $ */

/*
 * Copyright (c) 2001 Alistair G. Crooks.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by Alistair G. Crooks.
 * 4. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior written
 *    permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS
 * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <digest-types.h>

#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 2001 \
	        The NetBSD Foundation, Inc.  All rights reserved.");
__RCSID("$NetBSD: digest.c,v 1.1.1.1 2002/09/19 10:44:28 agc Exp $");
#endif


#ifdef HAVE_ERRNO_H
#include <errno.h>
#endif
#ifdef HAVE_LOCALE_H
#include <locale.h>
#endif
#include <md5.h>
#include <rmd160.h>
#include <sha1.h>
#include <sha2.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

/* perform an md5 digest, and print the results if successful */
static int
md5_digest_file(char *fn)
{
	MD5_CTX	m;
	char	in[BUFSIZ * 20];
	char	digest[33];
	int	cc;

	if (fn == NULL) {
		MD5Init(&m);
		while ((cc = read(STDIN_FILENO, in, sizeof(in))) > 0) {
			MD5Update(&m, (u_char *)in, (unsigned) cc);
		}
		(void) printf("%s\n", MD5End(&m, digest));
	} else {
		if (MD5File(fn, digest) == NULL) {
			return 0;
		}
		(void) printf("MD5 (%s) = %s\n", fn, digest);
	}
	return 1;
}

/* perform an sha1 digest, and print the results if successful */
static int
sha1_digest_file(char *fn)
{
	SHA1_CTX	sha;
	char		in[BUFSIZ * 20];
	char		digest[41];
	int		cc;

	if (fn == NULL) {
		SHA1Init(&sha);
		while ((cc = read(STDIN_FILENO, in, sizeof(in))) > 0) {
			SHA1Update(&sha, (u_char *)in, (unsigned) cc);
		}
		(void) printf("%s\n", SHA1End(&sha, digest));
	} else {
		if (SHA1File(fn, digest) == NULL) {
			return 0;
		}
		(void) printf("SHA1 (%s) = %s\n", fn, digest);
	}
	return 1;
}

/* perform an ripemd160 digest, and print the results if successful */
static int
rmd160_digest_file(char *fn)
{
	RMD160_CTX	rmd;
	char		in[BUFSIZ * 20];
	char		digest[41];
	int		cc;

	if (fn == NULL) {
		RMD160Init(&rmd);
		while ((cc = read(STDIN_FILENO, in, sizeof(in))) > 0) {
			RMD160Update(&rmd, (u_char *)in, (unsigned) cc);
		}
		(void) printf("%s\n", RMD160End(&rmd, digest));
	} else {
		if (RMD160File(fn, digest) == NULL) {
			return 0;
		}
		(void) printf("RMD160 (%s) = %s\n", fn, digest);
	}
	return 1;
}

/* perform a sha256 digest, and print the results if successful */
static int
sha256_digest_file(char *fn)
{
	SHA256_CTX	sha256;
	char		in[BUFSIZ * 20];
	char		digest[65];
	int		cc;

	if (fn == NULL) {
		SHA256_Init(&sha256);
		while ((cc = read(STDIN_FILENO, in, sizeof(in))) > 0) {
			SHA256_Update(&sha256, (u_char *)in, (unsigned) cc);
		}
		(void) printf("%s\n", SHA256_End(&sha256, digest));
	} else {
		if (SHA256_File(fn, digest) == NULL) {
			return 0;
		}
		(void) printf("SHA256 (%s) = %s\n", fn, digest);
	}
	return 1;
}

/* perform a sha384 digest, and print the results if successful */
static int
sha384_digest_file(char *fn)
{
	SHA384_CTX	sha384;
	char		in[BUFSIZ * 20];
	char		digest[97];
	int		cc;

	if (fn == NULL) {
		SHA384_Init(&sha384);
		while ((cc = read(STDIN_FILENO, in, sizeof(in))) > 0) {
			SHA384_Update(&sha384, (u_char *)in, (unsigned) cc);
		}
		(void) printf("%s\n", SHA384_End(&sha384, digest));
	} else {
		if (SHA384_File(fn, digest) == NULL) {
			return 0;
		}
		(void) printf("SHA384 (%s) = %s\n", fn, digest);
	}
	return 1;
}

/* perform a sha512 digest, and print the results if successful */
static int
sha512_digest_file(char *fn)
{
	SHA512_CTX	sha512;
	char		in[BUFSIZ * 20];
	char		digest[129];
	int		cc;

	if (fn == NULL) {
		SHA512_Init(&sha512);
		while ((cc = read(STDIN_FILENO, in, sizeof(in))) > 0) {
			SHA512_Update(&sha512, (u_char *)in, (unsigned) cc);
		}
		(void) printf("%s\n", SHA512_End(&sha512, digest));
	} else {
		if (SHA512_File(fn, digest) == NULL) {
			return 0;
		}
		(void) printf("SHA512 (%s) = %s\n", fn, digest);
	}
	return 1;
}

/* this struct defines a message digest algorithm */
typedef struct alg_t {
	const char     *name;			/* algorithm name */
	int		(*func)(char *);	/* function to call */
} alg_t;

/* list of supported message digest algorithms */
static alg_t algorithms[] = {
	{ "md5",	md5_digest_file		},
	{ "rmd160",	rmd160_digest_file	},
	{ "sha1",	sha1_digest_file	},
	{ "sha256",	sha256_digest_file	},
	{ "sha384",	sha384_digest_file	},
	{ "sha512",	sha512_digest_file	},
	{ NULL	}
};

/* find an algorithm, given a name */
static alg_t *
find_algorithm(const char *a)
{
	alg_t	*alg;

	for (alg = algorithms ; alg->name && strcasecmp(alg->name, a) != 0 ; alg++) {
	}
	return (alg->name) ? alg : NULL;
}

int
main(int argc, char **argv)
{
	alg_t  *alg;
	int	rval;
	int	i;

#ifdef HAVE_SETLOCALE
	(void) setlocale(LC_ALL, "");
#endif
	while ((i = getopt(argc, argv, "V")) != -1) {
		switch(i) {
		case 'V':
			printf("%s\n", VERSION);
			return EXIT_SUCCESS;
		}
	}
	if (argc == optind) {
		(void) fprintf(stderr, "Usage: %s algorithm [file...]\n", *argv);
		return EXIT_FAILURE;
	}
	if ((alg = find_algorithm(argv[optind])) == NULL) {
		(void) fprintf(stderr, "No such algorithm `%s'\n", argv[optind]);
		exit(EXIT_FAILURE);
	}
	rval = EXIT_SUCCESS;
	if (argc == optind + 1) {
		if (!(*alg->func)(NULL)) {
			(void) fprintf(stderr, "stdin\n");
			rval = EXIT_FAILURE;
		}
	} else {
		for (i = optind + 1 ; i < argc ; i++) {
			if (!(*alg->func)(argv[i])) {
				(void) fprintf(stderr, "%s\n", argv[i]);
				rval = EXIT_FAILURE;
			}
		}
	}
	return rval;
}
